From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ricardo Salveti de Araujo <ricardo.salveti@canonical.com>
Date: Thu, 31 Oct 2013 12:19:19 +0200
Subject: [PATCH] (hybris) property_service.c: adding support for getprop and
 listprop
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Change-Id: Ie5fbdb55c48038ce8250f27500623b3b81cc5cd1
Signed-off-by: Jani Monoses <jani@ubuntu.com>
Signed-off-by: Ricardo Salveti de Araujo <ricardo.salveti@canonical.com>
Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Conflicts:
	init/property_service.c

Conflicts:
	init/property_service.cpp
---
 init/init.cpp             |  2 +-
 init/property_service.cpp | 61 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 62 insertions(+), 1 deletion(-)

diff --git a/init/init.cpp b/init/init.cpp
index 79f2f235f238b0d4396fec5514a1a8eebee60847..94fc72060c4a18a5e3a7702cc5c474054855f5fc 100644
--- a/init/init.cpp
+++ b/init/init.cpp
@@ -1099,7 +1099,7 @@ int SecondStageMain(int argc, char** argv) {
     am.QueueBuiltinAction(queue_property_triggers_action, "queue_property_triggers");
 
     /* run all device triggers based on current state of device nodes in /dev */
-    am.QueueBuiltinAction(queue_device_triggers_action, "queue_device_triggers");
+    //m.QueueBuiltinAction(queue_device_triggers_action, "queue_device_triggers");
 
     /* Run actions when all boot up is done and init is ready */
     am.QueueEventTrigger("ready");
diff --git a/init/property_service.cpp b/init/property_service.cpp
index 157fc709c6be5b7733a4b5399dabf87082d7ba67..c7f880c02f6d5159a9b38e43e121790bb2ee6b54 100644
--- a/init/property_service.cpp
+++ b/init/property_service.cpp
@@ -198,6 +198,12 @@ void NotifyPropertyChange(const std::string& name, const std::string& value) {
         PropertyChanged(name, value);
     }
 }
+// Legacy definition
+typedef struct prop_msg_s {
+    unsigned cmd;
+    char name[PROP_NAME_MAX];
+    char value[PROP_VALUE_MAX];
+} prop_msg_t;
 
 class AsyncRestorecon {
   public:
@@ -296,6 +302,17 @@ class SocketConnection {
 
     [[nodiscard]] int Release() { return socket_.release(); }
 
+    bool SendLegacyPropMsg(std::string name, std::string value) {
+        int result = 0;
+        prop_msg_t prop_msg;
+        memset(&prop_msg, 0, sizeof(prop_msg_t));
+        prop_msg.cmd = 0;
+        strlcpy(prop_msg.name, name.c_str(), PROP_NAME_MAX);
+        strlcpy(prop_msg.value, value.c_str(), PROP_VALUE_MAX);
+        result = TEMP_FAILURE_RETRY(send(socket_.get(), &prop_msg, sizeof(prop_msg_t), 0));
+        return (result == sizeof(prop_msg_t));
+    }
+
     const ucred& cred() { return cred_; }
 
     SocketConnection& operator=(SocketConnection&&) = default;
@@ -597,6 +614,9 @@ uint32_t HandlePropertySetNoSocket(const std::string& name, const std::string& v
 static void handle_property_set_fd(int fd) {
     static constexpr uint32_t kDefaultSocketTimeout = 2000; /* ms */
 
+    char rproperty[PROP_VALUE_MAX];
+    const prop_info *pi;
+
     int s = accept4(fd, nullptr, nullptr, SOCK_CLOEXEC);
     if (s == -1) {
         return;
@@ -683,7 +703,48 @@ static void handle_property_set_fd(int fd) {
         socket.SendUint32(*result);
         break;
       }
+    case PROP_MSG_GETPROP: {
+        int ret;
+        char prop_name[PROP_NAME_MAX];
+        char prop_value[PROP_VALUE_MAX];
+
+        /* Need to receive both since libhybris sends prop_msg_t type package. */
+        if (!socket.RecvChars(prop_name, PROP_NAME_MAX, &timeout_ms) ||
+            !socket.RecvChars(prop_value, PROP_VALUE_MAX, &timeout_ms)) {
+            PLOG(ERROR) << "sys_prop(PROP_MSG_GETPROP): error while reading name from the socket";
+            return;
+        }
+
+        prop_name[PROP_NAME_MAX-1] = 0;
+        prop_value[PROP_VALUE_MAX-1] = 0;
 
+        if (!IsLegalPropertyName(prop_name)) {
+            PLOG(ERROR) << "sys_prop(PROP_MSG_GETPROP): illegal property name. Got: " << prop_name;
+            return;
+        }
+
+        /* If we have a value, copy it over, otherwise returns the default */
+        ret = __system_property_get(prop_name, rproperty);
+        if (ret) {
+            strlcpy(prop_value, rproperty, sizeof(prop_value));
+        }
+
+        /* Send the property value back */
+        //TEMP_FAILURE_RETRY(send(s, &msg, sizeof(msg), 0));
+        socket.SendLegacyPropMsg(prop_name, prop_value);
+        break;
+      }
+    case PROP_MSG_LISTPROP: {
+        char prop_name[PROP_NAME_MAX];
+        char prop_value[PROP_VALUE_MAX];
+        for(int n = 0; (pi = __system_property_find_nth(n)); n++) {
+            prop_name[0] = prop_value[0] = 0;
+            __system_property_read(pi, prop_name, prop_value);
+            //TEMP_FAILURE_RETRY(send(s, &msg, sizeof(msg), 0));
+            socket.SendLegacyPropMsg(prop_name, prop_value);
+        }
+        break;
+      }
     default:
         LOG(ERROR) << "sys_prop: invalid command " << cmd;
         socket.SendUint32(PROP_ERROR_INVALID_CMD);
